### PR TITLE
Improvements for Java support

### PR DESCRIPTION
**Pull request by bleshik**
Workarounded Java type check problem https://github.com/ensime/ensime-server/issues/1616
Added Syntastic support for Java
Added autocommands for *.java

### CODE REVIEW

File: syntax_checkers/java/ensime.vim [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+function! SyntaxCheckers_java_ensime_GetLocList() dict
+    if exists('b:ensime_notes')
+        return b:ensime_notes
+    else
+        return []
+    endif
+endfunction
+
+function! SyntaxCheckers_java_ensime_GetHighlightRegex(error)
```
Comment (jvican) [inline syntax_checkers/java/ensime.vim]: Why is this function required?

File: ensime_shared/typecheck.py [hunk 2]
```diff
@@ -13,6 +13,14 @@ def buffer_typechecks(self, call_id, payload):
...
             for note in payload['notes']:
                 self.buffered_notes.append(note)
+            self.editor.display_notes(self.buffered_notes)
+
+    def buffer_typechecks_and_display(self, call_id, payload):
+        """Adds typecheck events to the buffer, and displays them right away.
+        This is currently used as a workaround for issue https://github.com/ensime/ensime-server/issues/1616
+        """
+        self.buffer_typechecks(call_id, payload)
+        self.editor.display_notes(self.buffered_notes)
```
Comment (jvican) [inline ensime_shared/typecheck.py]: Why is this line required? `buffer_typechecks` already displays notes. I think L16 or L23 could go away!

Comment (bleshik) [inline ensime_shared/typecheck.py]: Oops, L16 should be deleted

File: syntax_checkers/java/ensime.vim (hunk 1 above)
Comment (bleshik) [inline syntax_checkers/java/ensime.vim]: Well, in order to make errors look just like it currently looks for scala, it is required (see SyntaxCheckers_scala_ensime_GetHighlightRegex). I could extract duplicated code in a separate function, but there a few lines. Should I still do it?

Comment (jvican) [inline syntax_checkers/java/ensime.vim]: Nope, don't worry :smile:.

File: ensime_shared/typecheck.py [hunk 3]
```diff
@@ -14,6 +14,13 @@ def buffer_typechecks(self, call_id, payload):
             for note in payload['notes']:
                 self.buffered_notes.append(note)
 
+    def buffer_typechecks_and_display(self, call_id, payload):
+        """Adds typecheck events to the buffer, and displays them right away.
+        This is currently used as a workaround for issue https://github.com/ensime/ensime-server/issues/1616
```
Comment (ktonga) [inline ensime_shared/typecheck.py]: Could you please create an issue to track this? if it is a temporary "hack" it should be nice to keep an eye on it and do the right thing once the server issue gets fixed.

Comment (bleshik) [inline ensime_shared/typecheck.py]: Done https://github.com/ensime/ensime-vim/issues/361

File: plugin/ensime.vim [hunk 4]
```diff
@@ -19,11 +19,11 @@ endif
...
-    autocmd VimLeave *.scala call ensime#au_vim_leave(expand("<afile>"))
-    autocmd VimEnter *.scala call ensime#au_vim_enter(expand("<afile>"))
-    autocmd BufLeave *.scala call ensime#au_buf_leave(expand("<afile>"))
-    autocmd CursorHold *.scala call ensime#au_cursor_hold(expand("<afile>"))
-    autocmd CursorMoved *.scala call ensime#au_cursor_moved(expand("<afile>"))
+    autocmd VimLeave *.java,*.scala call ensime#au_vim_leave(expand("<afile>"))
+    autocmd VimEnter *.java,*.scala call ensime#au_vim_enter(expand("<afile>"))
+    autocmd BufLeave *.java,*.scala call ensime#au_buf_leave(expand("<afile>"))
+    autocmd CursorHold *.java,*.scala call ensime#au_cursor_hold(expand("<afile>"))
+    autocmd CursorMoved *.java,*.scala call ensime#au_cursor_moved(expand("<afile>"))
```
Comment (ches) [inline plugin/ensime.vim]: This needs to be done for Neovim as well in the `rplugin` file.

Comment (ktonga) [inline plugin/ensime.vim]: oh, you're right, totally forgot that. TBH it's a PITA, it would be nice to have at least a test that verifies both vim plugin and nvim remote plugin define the same commands and autocommands.
